tests/drift_kinetic_push/drift_kinetic_grid_ex3.cpp

```diff
@@ -0,0 +1,330 @@
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Max B field error:     %.8e\n", stats.max_field_error_B));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Max gradB field error: %.8e\n", stats.max_field_error_gradB));
+
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "\nTrajectory comparison:\n"));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Max position error:    %.8e\n", stats.max_position_error));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Final pos analytical:  (%.6e %.6e %.6e)\n",
+           REP3_A(stats.final_position_analytical)));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Final pos grid:        (%.6e %.6e %.6e)\n",
+           REP3_A(stats.final_position_grid)));
+
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "\nIteration statistics:\n"));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Total iterations (analytical): %d\n", stats.total_iterations_analytical));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Total iterations (grid):       %d\n", stats.total_iterations_grid));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Avg iterations (analytical):   %.2f\n",
+           (PetscReal)stats.total_iterations_analytical / stats.total_steps));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Avg iterations (grid):         %.2f\n",
+           (PetscReal)stats.total_iterations_grid / stats.total_steps));
+
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "\nConstraint violations:\n"));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Boundary violations:    %d\n", stats.boundary_violations));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Cell size violations:   %d\n", stats.cell_size_violations));
+  PetscCall(PetscPrintf(PETSC_COMM_WORLD, "  Max displacement/step:  %.8e\n", stats.max_displacement_per_step));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+int main(int argc, char** argv)
+{
+  PetscFunctionBeginUser;
+  PetscCall(PetscInitialize(&argc, &argv, nullptr, help));
+
+  PetscReal omega_dt;
+  PetscCall(get_omega_dt(omega_dt));
+
+  dt = omega_dt / B0.length();
+  geom_nt = 100;
+  diagnose_period = geom_nt / 1;
+  geom_nx = 20;
+  geom_ny = 20;
+  geom_nz = 20;
+  dx = 1.;
+  dy = 1.;
+  dz = 1.;
+
+
+  World::set_geometry(
+    geom_nx, geom_ny, geom_nz, geom_nt,
+    dx, dy, dz, dt,
+    diagnose_period);
+
+  World world;
+  PetscCall(world.initialize());
+
+  Vec E_vec, B_vec, gradB_vec;
+  PetscCall(DMCreateGlobalVector(world.da, &E_vec));
+  PetscCall(DMCreateGlobalVector(world.da, &B_vec));
+  PetscCall(DMCreateGlobalVector(world.da, &gradB_vec));
+
+  PetscCall(initialize_grid_fields(world.da, {dx,dy,dz}, E_vec, B_vec, gradB_vec));
+
+  Vector3R*** E_arr, ***B_arr, ***gradB_arr;
+  PetscCall(DMDAVecGetArrayRead(world.da, E_vec, &E_arr));
+  PetscCall(DMDAVecGetArrayRead(world.da, B_vec, &B_arr));
+  PetscCall(DMDAVecGetArrayRead(world.da, gradB_vec, &gradB_arr));
+
+  std::unique_ptr<DriftKineticEsirkepov> esirkepov =
+    std::make_unique<DriftKineticEsirkepov>(E_arr, B_arr, nullptr, gradB_arr);
```

Comment:
Use of pointers here is an overkill, simply

```suggestion
  DriftKineticEsirkepov esirkepov(E_arr, B_arr, nullptr, gradB_arr);
```